### PR TITLE
feat: use wordpress for static pages  

### LINKED ISSUE #7983: Experiment: Try to integrate a wordpress page into the main site
To tackle:
- #1818

One way would be to use a real CMS and integrate page contents in the website. It should feel natural (no change of top banner, etc.)

A natural candidate CMS is wordpress as:
- the UI is well known by a lot of people
- we already use it for the blog
- it has good support for translation and even crowdin integration

But we are unsure if it's easy to integrate the content in another website. It should feel really natural for editors to preview the content on the wordpress and see it published.

(Note: we might also do other experiments with other software, this ticket is not to debate about it)

A first step should be to try to integrate this page: https://blog.openfoodfacts.org/en/press-release-new-mobile-app to the website.

### LINKED ISSUE #10539: Enable handling redirects in openfoodfacts-web
### Problem
In openfoodfacts-web, symlinks enable pointing some urls to same content.

Sometime it's better to have redirects, for url that changed for example.

We have a lot of redirects right now directly in the nginx configuration, (see https://github.com/openfoodfacts/search-a-licious/)

It would be better if all this can be handled naturally in openfoodfacts-web.

### Proposed solution

Handle this case by looking at files in openfoodfacts-web.

I think the best would be to have a file `lang/texts/<lc>/<location>.redirect` and inside put a URI to redirect to.

As we do for html files, at startup, we should gather those redirects, so that in display content function we are able to issue the right redirect.

### PR DESCRIPTION
**Pull request by 4nt0ineB**
### What

Experience of using Wordpress as a CMS for static pages. Page content is retrieved from the API. 
The goal is to render the page as close as it is displayed in the Gutenberg editor.

- [x] WPML set up
- [x] [Procedure to migrate off-web pages to wordpress](https://wiki.openfoodfacts.org/Open_Food_Facts_Contents)
- [x] #10539
- [ ] Crowdin integration

### Screenshot
Peek 2024-07-25 16-25

### Related issue(s) and discussion

- Fixes #7983
- Fixes #10539

### CODE REVIEW

File: html/css/wp.css
Comment (4nt0ineB) [inline html/css/wp.css:1]: The issue is that, while the html content retrieved contains the styles and classes added on the tags by Gutenberg , the css is not accessible. I add to find and combine by exploring F12. Only the necessary styles are loaded depending on the block types we used in the page.

**Comment by stephanegigandet** [comment]
Looks great to me, thank you!

**Comment by 4nt0ineB** [comment]
> First thanks for this great work !
> 
> I would prefer the cache to be handled differently, I think this way is problematic.
> 
> @4nt0ineB it would be important also to write a procedure on how to migrate from openfoodfacts-web to wordpress content progressively (how to replace a page). I think it could go with implementing #10539 (so basically, you would create page in wordpress and publish, then put a redirect to content/new-page in openfoodfacts-web).

Not sure if I understand correctly.

File: lib/ProductOpener/CMS.pm [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+The slug of the page to fetch: 
+e.g. 'my-test-page' or 'journees-open-food-facts-2024-reviennent-en-septembre-a-paris'
+
+=back
+
+=cut 
+
+sub wp_get_page_from_slug($lc, $slug) {
+
+	wp_update_pages_metadata_cache();
```
[the remark this answers is not in the document]
Comment (4nt0ineB) [reply lib/ProductOpener/CMS.pm]: > I'm not really found of this kind of update called silently within a request.
> 
> Also, here, if it take say, 1s, we may have the work done more than once in parallel, and you didn't protect access to it, so result is eventually unpredictable… (hitting two pages in same second is far from an exception I think).
> 
> Ideally I would have put this cache information in memcached, initialized it at startup if it's not in memcached (in apache startup), and just provided a url to refresh it on demand.
> 
> I think we can then either find a plugin on wordpress to call this url on page change (some webhook), or call it with a timer.

Damn I didn't thought about it at all, that it was parallel. I changed  the way of caching the page metadata, as you proposed. 
The content is currently not cached and when we need to display it, we get it by calling the WordPress API, every time. Do we want to cache the page content as well (which imply each translation) ?

[the remark this answers is not in the document]
Comment (alexgarel) [reply lib/ProductOpener/CMS.pm]: No need to store page content yet, maybe :-)

**Comment by teolemon** [comment]
Would it be possible to have the WordPress id and/or slug and/or custom properties exposed to ProductOpener so that we can create deep links like: "Help translate this article on Crowdin" or "Help translate the 2 associated visuals on Canva".
page_wp_id:17
or crowdin_page_wp_id:345
page_visual_1_translate_url:http://
page_visual_2_translate_url:http://

**Comment by 4nt0ineB** [comment]
> Would it be possible to have the WordPress id and/or slug and/or custom properties exposed to ProductOpener so that we can create deep links like: "Help translate this article on Crowdin" or "Help translate the 2 associated visuals on Canva". page_wp_id:17 or crowdin_page_wp_id:345 page_visual_1_translate_url:http:// page_visual_2_translate_url:http://

I haven't looked into Crowdin integration yet. But if a crowdin_id field is stored and exposed in the WPML api then I guess its possible.

**Comment by 4nt0ineB** [comment]
> I think what's missing in the PR is how we handle the different URLs slugs that are translated.
> 
> e.g. if we create a page "A new page", we want it to be available at world.openfoodfacts.org/content/a-new-page but also fr.openfoodfacts.org/content/a-new-page
> 
> when a new French translation is added, we want it to be available at fr.openfoodfacts.org/content/une-nouvelle-page , and fr.openfoodfacts.org/content/a-new-page should redirect to it.
> 
> but we shouldn't have to edit the .redirect file (as described here: https://wiki.openfoodfacts.org/Open_Food_Facts_Contents ) each time someone adds a translation..
> 
> am I missing something?
> 
> Ideally PO should find out from wordpress which pages are available, which are translated and what are their urls. Then we use that in the routing and displaying of the translated content if available, and the English content if not.
> 
> We could then automatically replace things like <content_url:a-new-page> to https://fr.openfoodfacts.org/content/une-nouvelle-page

Yes what I did is that PO already fetch all the available translations. I think the redirection was done for SEO purposes.

**Comment by john-gom** [comment]
Sorry for coming late to this but should we consider loading the wordpress content from the browser so that we are reducing the amount of traffic that flows through ProductOpener?

**Comment by alexgarel** [comment]
> Sorry for coming late to this but should we consider loading the wordpress content from the browser so that we are reducing the amount of traffic that flows through ProductOpener?

Yes it was considered, but we didn't want to maintain a specific layout on wordpress, and we wanted to have the menu and so on…

**Comment by john-gom** [comment]
> > Sorry for coming late to this but should we consider loading the wordpress content from the browser so that we are reducing the amount of traffic that flows through ProductOpener?
> 
> Yes it was considered, but we didn't want to maintain a specific layout on wordpress, and we wanted to have the menu and so on…

You can use module federation with a micro frontend to achieve this

**Comment by alexgarel** [comment]
Closing in favour of #10992

**Pull request closed without merging**